### PR TITLE
Add MiniMax Music 3

### PR DESCRIPTION
**Pull request by apolinario**
Adds MiniMax Music 3 (lyrics + music description → complete songs up to 5 minutes, 44.1 kHz stereo). An 8B Qwen3 autoregressive stage predicts per-frame audio codes; its hidden states condition a 2.4B flow-matching transformer that produces Flow-VAE latents in overlapping chunks; a DAC-style decoder renders the waveform. Structurally close to Ace-Step (DiT + pipeline-local submodels) and AudioLDM2 (language model as a pipeline component).

New: `MiniMaxMusic3Transformer1DModel`, `MiniMaxMusic3ConditionEncoder` (in `models/condition_embedders/`), `MiniMaxMusic3RVQDepthDecoder`, `MiniMaxMusic3Vocoder`, and `MiniMaxMusic3ModularPipeline` (modular blocks: text encode → AR semantic generation → chunked flow-matching with guider-abstracted CFG → vocode/stitch), plus conversion script, docs, and model- and modular-pipeline tests. The scheduler is the existing `FlowMatchEulerDiscreteScheduler(num_train_timesteps=1, shift=1.0, invert_sigmas=True)`, which reproduces the reference Euler loop exactly.

Usage:

```python
import soundfile as sf
import torch
from diffusers import ModularPipeline

pipe = ModularPipeline.from_pretrained("MiniMaxAI/MiniMax-Music3")
pipe.load_components(dtype=torch.bfloat16)
pipe.to("cuda")

lyrics = """[verse]
Morning light filtering through the pine
Every quiet street is yours and mine
[chorus]
Softly the world begins to breathe"""

prompt = (
    "Genre: acoustic pop. BPM: 96. Key: C major. Warm and intimate, building gently into the chorus. "
    "Vocals: soft female lead, close and breathy, light stacked harmonies in the chorus. "
    "Arrangement: fingerpicked guitar and soft piano; brushed drums and upright bass enter in the chorus."
)

audio = pipe(
    prompt=prompt,
    lyrics=lyrics,
    audio_duration=60.0,
    generator=torch.Generator("cuda").manual_seed(7),
    output="audios",
)[0]

sf.write("song.wav", audio.T.float().cpu().numpy(), pipe.sampling_rate)
```

Parity vs the reference implementation: every converted component matches bitwise on CPU/fp32; full 30-step chunked generation matches at 57.8 dB SNR on GPU (residual comes from the fused→split QKV projections).

Weights: final home will be `MiniMaxAI/MiniMax-Music3` when the model is out

Some choices I made that would be great to double check:
1. One `generator` drives the whole pipeline - the reference derives a separate seed per stage and per chunk, imo it's fine to keep the diffusers convention here.
2. Output is the vocoder's native 44.1 kHz; the reference server post-resamples to 32 kHz - I don't see a reason to downsample something higher quality, but let me know if following the original would be fundamental here
3. Generation is single-sample (`prompt`/`lyrics` are strings; the AR stage is internally batch-2 for CFG). If we want to optimize batched generation I think it may need an AR-loop redesign, which I would defer to a follow-up if someone would be interested?

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_minimax_music3.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+from ...utils.torch_utils import lru_cache_unless_export
+from ..attention import AttentionModuleMixin
+from ..attention_dispatch import dispatch_attention_fn
+from ..embeddings import TimestepEmbedding
+from ..modeling_utils import ModelMixin
+
+
+@dataclass
+class MiniMaxMusic3TransformerOutput(BaseOutput):
+    sample: torch.Tensor
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_minimax_music3.py]: Is it not possible to use the `Transformer2DModelOutput` here?

File: src/diffusers/models/transformers/transformer_minimax_music3.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    @lru_cache_unless_export(maxsize=32)
+    def _build(self, seq_len: int, device: torch.device) -> Tuple[torch.Tensor, torch.Tensor]:
+        inv_freq = 1.0 / (self.theta ** (torch.arange(0, self.rotary_dim, 2, device=device).float() / self.rotary_dim))
+        steps = torch.arange(seq_len, device=device, dtype=torch.float32)
+        freqs = torch.outer(steps, inv_freq)
+        freqs = torch.cat((freqs, freqs), dim=-1)
+        return freqs.cos().contiguous(), freqs.sin().contiguous()
+
+    def forward(self, seq_len: int, device: torch.device) -> Tuple[torch.Tensor, torch.Tensor]:
+        return self._build(seq_len, device)
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_minimax_music3.py]: Consider folding into `forward()` so that things stay explicitly inline.

File: src/diffusers/models/transformers/transformer_minimax_music3.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    def __init__(self, dim: int, heads: int, head_dim: int, processor: Optional[MiniMaxMusic3AttnProcessor] = None):
+        super().__init__()
+        self.heads = heads
+        self.head_dim = head_dim
+        self.inner_dim = heads * head_dim
+        self.to_q = nn.Linear(dim, self.inner_dim, bias=False)
+        self.to_k = nn.Linear(dim, self.inner_dim, bias=False)
+        self.to_v = nn.Linear(dim, self.inner_dim, bias=False)
+        self.to_out = nn.ModuleList([nn.Linear(self.inner_dim, dim, bias=False), nn.Dropout(0.0)])
+        self.set_processor(processor or MiniMaxMusic3AttnProcessor())
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_minimax_music3.py]:
```suggestion
        if processor is None:
            processor = self._default_processor_cls()
        self.set_processor(processor)
```

File: src/diffusers/pipelines/minimax_music3/pipeline_minimax_music3.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+        # trailing latents as an overlap prompt, then cropped so the kept spans tile the full song.
+        chunk_starts = [0] if num_frames <= _CHUNK_FRAMES else list(range(0, num_frames - _CHUNK_HOP, _CHUNK_HOP))
+        self._num_timesteps = num_inference_steps * len(chunk_starts)
+
+        waveform_chunks = []
+        latent_chunks = []
+        previous_latent = None
+        previous_condition = None
+        global_step = 0
+        with self.progress_bar(total=self._num_timesteps) as progress_bar:
```
Comment (sayakpaul) [inline src/diffusers/pipelines/minimax_music3/pipeline_minimax_music3.py]: Why not do `self.num_inference_steps` or put `num_inference_steps * len(chunk_starts)` in a separate variable and use it here? This feels like an antipattern to me.

**Review by sayakpaul — COMMENTED** [review]
Thanks!

File: src/diffusers/models/transformers/transformer_minimax_music3.py (hunk 1 above)
Comment (apolinario) [reply src/diffusers/models/transformers/transformer_minimax_music3.py]: Done — now returns `Transformer2DModelOutput`.

File: src/diffusers/models/transformers/transformer_minimax_music3.py (hunk 2 above)
Comment (apolinario) [reply src/diffusers/models/transformers/transformer_minimax_music3.py]: Done — folded into `forward()` with the cache decorator on it directly.

File: src/diffusers/models/transformers/transformer_minimax_music3.py (hunk 3 above)
Comment (apolinario) [reply src/diffusers/models/transformers/transformer_minimax_music3.py]: Done (also applied to the depth decoder's attention).

File: src/diffusers/pipelines/minimax_music3/pipeline_minimax_music3.py (hunk 4 above)
Comment (apolinario) [reply src/diffusers/pipelines/minimax_music3/pipeline_minimax_music3.py]: Gone with the modular conversion — this file is removed; step accounting now lives in the loop blocks.

**Comment by apolinario** [comment]
Converted to a modular pipeline per @yiyixuxu's suggestion. Also applied Sayak's comments.

File: src/diffusers/modular_pipelines/minimax_music3/denoise.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    PipelineState,
+)
+from ..modular_pipeline_utils import ComponentSpec, InputParam, OutputParam
+from .before_denoise import _CHUNK_FRAMES
+from .modular_pipeline import MiniMaxMusic3ModularPipeline
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+# Neighboring windows share 172 latent frames; the previous window's carry spans latent frames [L - 344, L - 172).
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/modular_pipelines/minimax_music3/denoise.py]: Sure, Reworded in 1a258e3.

File: src/diffusers/modular_pipelines/minimax_music3/denoise.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        latents = block_state.latents
+        timesteps = block_state.timesteps
+        overlap = block_state.overlap
+
+        # The unconditional branch conditions on zeros, not on a re-encoded empty prompt.
+        guider_inputs = {
+            "encoder_hidden_states": (block_state.condition, torch.zeros_like(block_state.condition)),
+        }
+
+        with tqdm(total=block_state.num_inference_steps) as progress_bar:
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/modular_pipelines/minimax_music3/denoise.py]: Switched to a single pipeline-level `progress_bar` over `chunks × steps`, driven from the loop wrapper (1a258e3).

File: src/diffusers/modular_pipelines/minimax_music3/encoders.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+                description=(
+                    "Concatenated per-frame hidden states of shape `[1, frames, num_codebooks * hidden_size]` that "
+                    "condition the flow-matching stage."
+                ),
+            ),
+        ]
+
+    @staticmethod
+    def check_inputs(block_state):
+        if block_state.audio_duration <= 0:
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/modular_pipelines/minimax_music3/encoders.py:284]: Added a guard where `max_frames` is computed (frame rate is config-derived, so `check_inputs` can't see it) 1a258e3.

File: src/diffusers/models/__init__.py [hunk 8]
```diff
@@ -91,6 +93,7 @@
     _import_structure["transformers.dual_transformer_2d"] = ["DualTransformer2DModel"]
     _import_structure["transformers.hunyuan_transformer_2d"] = ["HunyuanDiT2DModel"]
     _import_structure["transformers.latte_transformer_3d"] = ["LatteTransformer3DModel"]
+    _import_structure["transformers.minimax_music3_rvq_depth_decoder"] = ["MiniMaxMusic3RVQDepthDecoder"]
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/models/__init__.py]: Fixed in 1a258e3.

File: src/diffusers/models/transformers/transformer_minimax_music3.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        Args:
+            hidden_states (`torch.Tensor` of shape `(batch, in_channels, length)`):
+                Noisy Flow-VAE latents.
+            timestep (`torch.Tensor` of shape `(batch,)`):
+                Flow-matching time in `[0, 1]`, where 0 is pure noise and 1 is data.
+            encoder_hidden_states (`torch.Tensor` of shape `(batch, length, condition_dim)`):
+                Frame-aligned conditioning from `MiniMaxMusic3ConditionEncoder`. Pass zeros for the unconditional
+                branch of classifier-free guidance.
+            return_dict (`bool`, defaults to `True`):
+                Whether to return a [`~models.transformers.transformer_minimax_music3.Transformer2DModelOutput`]
```
[the remark this answers is not in the document]
Comment (apolinario) [reply src/diffusers/models/transformers/transformer_minimax_music3.py]: Fixed in 1a258e3.

**Comment by apolinario** [comment]
dafe373 is a bug-fix for offloading. it fixes `ComponentsManager` CPU offloading: the AR step calls LM/depth-decoder submodules, which was bypassing the offload hook on the top-level forward. Applied the same manual `pre_forward` trigger minimax_h3 uses. Measured on A100: auto offload runs a generation in ~21 GB free (peak 17.3 GB); with `apply_group_offloading` on the language model it fits in 8 GB (peak 5.5 GB).

File: src/diffusers/modular_pipelines/minimax_music3/encoders.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+    offsets = (
+        torch.arange(components.num_codebooks - 1, device=frame_codes.device) * components.audio_vocab_size
+    ).unsqueeze(0)
+    extra = components.rvq_depth_decoder.audio_embeddings(frame_codes[:, 1:] + offsets).sum(dim=1, keepdim=True)
+    embeds = embeds + extra.to(embeds.dtype)
+    return embeds * components.num_codebooks**-0.5
+
+
+def _generate_depth_codes(
+    components: MiniMaxMusic3ModularPipeline,
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/minimax_music3/encoders.py]:
```suggestion
def _generate_depth_codes(
    language_model, 
    rvq_depth_decoder,
```
it's better to take the individual components as inputs

File: src/diffusers/modular_pipelines/minimax_music3/encoders.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+    threshold = torch.topk(values, top_k, dim=-1).values[..., -1, None]
+    values = values.masked_fill(values < threshold, -float("inf"))
+    probs = torch.nan_to_num(F.softmax(values, dim=-1), nan=0.0)
+    probs = probs / probs.sum(dim=-1, keepdim=True).clamp_min(1e-12)
+    # Sample on the generator's device so a CPU generator gives device-independent results (the diffusers convention).
+    sample_device = generator.device if generator is not None else probs.device
+    return torch.multinomial(probs.to(sample_device), 1, generator=generator).squeeze(-1).to(probs.device)
+
+
+def _embed_audio_frame(components: MiniMaxMusic3ModularPipeline, frame_codes: torch.Tensor) -> torch.Tensor:
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/minimax_music3/encoders.py]:
```suggestion
def _embed_audio_frame(
    language_model: Qwen3ForCausalLM, 
    rvq_depth_decoder: MiniMaxMusic3RVQDepthDecoder
    frame_codes: torch.Tensor) -> torch.Tensor:
```

File: src/diffusers/modular_pipelines/minimax_music3/encoders.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+        # The sampled code is repeated so the language-model feedback keeps the [conditional, unconditional] rows.
+        code = _sample_top_k(logits, generator).repeat(2)
+        codes.append(code)
+        if index < components.num_codebooks - 1:
+            embed = components.rvq_depth_decoder.audio_embeddings(code + (index - 1) * components.audio_vocab_size)
+            sequence.append(components.rvq_depth_decoder.projection(embed).unsqueeze(1))
+    return torch.stack(codes, dim=1), torch.cat(hidden_parts, dim=-1)
+
+
+class MiniMaxMusic3TextEncoderStep(ModularPipelineBlocks):
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/minimax_music3/encoders.py]: ohh it's actually an tokenize step, can we rename and then in the final assembled `MiniMaxMusic3Block`, merge it into the `MiniMaxMusic3SemanticGenerationStep`

File: src/diffusers/modular_pipelines/minimax_music3/modular_blocks_minimax_music3.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    """
+
+    block_classes = [
+        MiniMaxMusic3TextEncoderStep,
+        MiniMaxMusic3SemanticGenerationStep,
+        MiniMaxMusic3PrepareChunksStep,
+        MiniMaxMusic3ChunkDenoiseStep,
+        MiniMaxMusic3VocoderDecodeStep,
+    ]
+    block_names = ["text_encoder", "semantic_generator", "prepare_chunks", "denoise", "decode"]
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/minimax_music3/modular_blocks_minimax_music3.py]:
```suggestion
    block_names = ["semantic_generator", "denoise", "decode"]
```
I think each child block here should be ones that contain models and makes sense to be run standalone.e.g. user can use semantic_generator once and resue the output to denoise with different configs 


I think we can merge `MiniMaxMusic3TextEncoderStep` and `MiniMaxMusic3SemanticGenerationStep` into one sequential and then `prepare_chunks` +  `"denoise"` into the "MiniMaxMusic3CoreDenoiseStep" (same as MiniMaxH3 etc)

**Review by yiyixuxu — COMMENTED** [review]
thanks @apolinario i left some small comments

**Review by yiyixuxu — APPROVED** [review]
thanks a lot @apolinario !!
